index.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env python3
+
+"""
+Utility to rewrite all index.md files with the list of markdown articles
+discovered on disk. If an index.md file exist already, this utility
+conserve all lines prior to an "## Index" title (or the whole file if
+that title is missing).
+"""
+
+import dataclasses
+import itertools
+import subprocess as sp
+from collections import defaultdict
+from contextlib import suppress
+from datetime import datetime
+from functools import partial
+from operator import attrgetter
+from os import getenv
+from pathlib import Path
+from typing import List
+
+
+ROOT = Path(__file__).parent.resolve()
+GIT_PATH = getenv('GIT_PATH', '/usr/bin/git')
+
+
+def Tree():
+    return defaultdict(Tree)
+
+
+@dataclasses.dataclass
+class Article:
+    title: str
+    write_date: datetime
+    create_date: datetime
+    authors: List[str] = dataclasses.field(default_factory=list)
+    paths: List[Path] = dataclasses.field(default_factory=list)
+
+    @property
+    def name(self):
+        return self.paths[0].name
+
+    @property
+    def paths_parts(self):
+        # [/srv/git/Not-a-hub/langages/cpp/foo.md]
+        # => [["langages", "cpp"]]
+        return [path.parts[len(ROOT.parts):-1] for path in self.paths]
+
+    @property
+    def uris(self):
+        # [/srv/git/Not-a-hub/langages/cpp/foo.md]
+        # => ["/langages/cpp/foo.md"]]
+        return [str(path)[len(str(ROOT)):] for path in self.paths]
+
+    @property
+    def tags(self):
+        # [/srv/git/Not-a-hub/langages/cpp/foo.md]
+        # => {('langages', '/langages'), ('cpp', '/langages/cpp')}
+
+        def addsep(uri, part):
+            return f'{uri}/{part}'
+
+        to_uri = partial(itertools.accumulate, func=addsep)
+
+        return list(sorted({
+            (part, f'/{uri}')
+            for parts in self.paths_parts
+            for part, uri in zip(parts, to_uri(parts))
+        }))
+
+    def uri_for_root(self, root):
+        """
+        From all the paths of the current file, return one that starts
+        with ``root``.
+        """
+        return next(uri for uri in self.uris if uri.startswith(root))
+
+
+def extract_title(path):
+    """ Extract the first top level title of a markdown document """
+    assert path.suffix == '.md'
+
+    with open(path) as fd:
+        # lookahead the next line of every line during iteration
+        it1, it2 = itertools.tee(fd)
+        with suppress(StopIteration):
+            next(it2)
+        for current_line, next_line in itertools.zip_longest(it1, it2):
```

Comment:
est-ce qu'on pourrait pas tout simplement dire que la première ligne du fichier c'est le titre de l'article ? (en vérifiant tout de même le # ou les = evidemment)
qu'est ce qu'on pourrait avoir d'autre avant ?
parce que la j'ai l'impression qu'on va créer des itérateurs pour rien dans 99% des cas

Reply:
oui ça va créer un itérateur pour rien à 99% du temps mais ainsi c'est robuste :D